### PR TITLE
[google_sign_in] Add iOS unit tests

### PR DESCRIPTION
**Pull request by jmagman**
- Add google_sign_in iOS unit tests to validate we don't break (much) in preparation for https://github.com/flutter/flutter/issues/86436, which included SDK removals and signature changes without deprecations. Native code plugin coverage increased from 41.4% to 78.4%.
- Add `-[FLTGoogleSignInPlugin initWithSignIn:]` initializer to allow easy injection of `GIDSignIn` into unit tests.  Expose in a new `google_sign_in.Test` explicit clang module.  Include new modulemap and umbrella header to support this.
- Add lightweight generics to the plugin.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

File: packages/google_sign_in/google_sign_in/example/ios/Flutter/AppFrameworkInfo.plist [hunk 1]
```diff
@@ -25,6 +25,6 @@
     <string>arm64</string>
   </array>
   <key>MinimumOSVersion</key>
-  <string>8.0</string>
+  <string>9.0</string>
```
Comment (jmagman) [inline packages/google_sign_in/google_sign_in/example/ios/Flutter/AppFrameworkInfo.plist:28]: Automatic from https://github.com/flutter/flutter/pull/85174

File: packages/google_sign_in/google_sign_in/example/ios/Runner.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -539,7 +539,7 @@
 				GCC_WARN_UNINITIALIZED_AUTOS = YES_AGGRESSIVE;
 				GCC_WARN_UNUSED_FUNCTION = YES;
 				GCC_WARN_UNUSED_VARIABLE = YES;
-				IPHONEOS_DEPLOYMENT_TARGET = 8.0;
+				IPHONEOS_DEPLOYMENT_TARGET = 9.0;
```
Comment (jmagman) [inline packages/google_sign_in/google_sign_in/example/ios/Runner.xcodeproj/project.pbxproj:542]: Automatic from https://github.com/flutter/flutter/pull/85174

File: packages/google_sign_in/google_sign_in/example/ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme [hunk 3]
```diff
@@ -26,7 +26,8 @@
       buildConfiguration = "Debug"
       selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
       selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
-      shouldUseLaunchSchemeArgsEnv = "YES">
+      shouldUseLaunchSchemeArgsEnv = "YES"
+      codeCoverageEnabled = "YES">
```
Comment (jmagman) [inline packages/google_sign_in/google_sign_in/example/ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme:30]: Turn on code coverage.

File: packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m [hunk 4]
```diff
@@ -16,7 +16,7 @@ @interface FLTGoogleSignInPluginTest : XCTestCase
 @property(strong, nonatomic) NSObject<FlutterBinaryMessenger> *mockBinaryMessenger;
 @property(strong, nonatomic) NSObject<FlutterPluginRegistrar> *mockPluginRegistrar;
 @property(strong, nonatomic) FLTGoogleSignInPlugin *plugin;
-@property(strong, nonatomic) GIDSignIn *mockSharedInstance;
+@property(strong, nonatomic) id mockSharedInstance;
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m:19]: Can we give this variable a clearer name, especially since it no longer has a type?

File: packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m [hunk 5]
```diff
@@ -26,62 +26,406 @@ - (void)setUp {
   [super setUp];
   self.mockBinaryMessenger = OCMProtocolMock(@protocol(FlutterBinaryMessenger));
   self.mockPluginRegistrar = OCMProtocolMock(@protocol(FlutterPluginRegistrar));
-  self.mockSharedInstance = [OCMockObject partialMockForObject:[GIDSignIn sharedInstance]];
+
+  id mockSharedInstance = OCMClassMock([GIDSignIn class]);
+  OCMStub([mockSharedInstance sharedInstance]).andReturn(mockSharedInstance);
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m:31]: Any chance we could adjust the code to accept an injectable GIDSignIn instance instead of using class mocking?

File: packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m [hunk 6]
```diff
@@ -26,62 +26,406 @@ - (void)setUp {
...
+  OCMVerify([self.mockSharedInstance signOut]);
+}
+
+- (void)testDisconnect {
+  FlutterMethodCall *methodCall = [FlutterMethodCall methodCallWithMethodName:@"disconnect"
+                                                                    arguments:nil];
+
+  [self.plugin handleMethodCall:methodCall
+                         result:^(id result){
+                         }];
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m:77]: Don't we want to assert that it returned a (non-error) result?

File: packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m [hunk 7]
```diff
@@ -26,62 +26,406 @@ - (void)setUp {
...
+
+- (void)testSignInSilently {
+  OCMExpect([self.mockSharedInstance restorePreviousSignIn]);
+
+  FlutterMethodCall *methodCall = [FlutterMethodCall methodCallWithMethodName:@"signInSilently"
+                                                                    arguments:nil];
+
+  [self.plugin handleMethodCall:methodCall
+                         result:^(id result){
+                         }];
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m:203]: Same question here. (And all following cases that don't assert anything about the return value)

File: packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m [hunk 8]
```diff
@@ -26,62 +26,406 @@ - (void)setUp {
...
+                           XCTAssertEqualObjects(result.code, @"sign_in_canceled");
+                           XCTAssertEqualObjects(result.message, kGIDSignInErrorDomain);
+                           [expectation fulfill];
+                         }];
+  [self waitForExpectationsWithTimeout:5.0 handler:nil];
+}
+
+- (void)testGetTokensURLError {
+  id mockUser = OCMClassMock([GIDGoogleUser class]);
+  OCMStub([self.mockSharedInstance currentUser]).andReturn(mockUser);
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m:340]: Don't these need stopMocking calls?

File: packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin.m [hunk 9]
```diff
@@ -76,7 +76,8 @@ - (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result
       NSString *path = [[NSBundle mainBundle] pathForResource:@"GoogleService-Info"
                                                        ofType:@"plist"];
       if (path) {
-        NSMutableDictionary *plist = [[NSMutableDictionary alloc] initWithContentsOfFile:path];
+        NSMutableDictionary<NSString *, NSString *> *plist =
+            [[NSMutableDictionary alloc] initWithContentsOfFile:path];
         BOOL hasDynamicClientId =
             [[call.arguments valueForKey:@"clientId"] isKindOfClass:[NSString class]];
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin.m:82]: Maybe use that new-fangled `[]` access here while you are modernizing :)

(We could consider doing a sweep of the repo for old-style dictionary and array reads and writes and updating them, like I did for literals.)

**Review by stuartmorgan — COMMENTED** [review]
Nice :)

The description says "Turn on code coverage", but I'm not seeing project changes other than deployment target.

File: packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m (hunk 6 above)
Comment (jmagman) [reply packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m:77]: The tests that don't check the result are doing do because the result callback happens from a `GIDSignIn.delegate` method.  
https://github.com/flutter/plugins/blob/b548ae18646d1725a8e1f004c26bf42e4d362aa8/packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin.m#L132-L135
sends the response in the `GIDSignInDelegate` call.
https://github.com/flutter/plugins/blob/b548ae18646d1725a8e1f004c26bf42e4d362aa8/packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin.m#L246-L250
which means I'd have to mock out the delegate calls.  
GoogleSignIn 6.0 (https://github.com/flutter/flutter/issues/86436) removed this delegate and made it a `disconnectWithCallback:` so the above code will become:
```obj-c
    [GIDSignIn.sharedInstance disconnectWithCallback:^(NSError * _Nullable error) {
      [self respondWithAccount:@{} result:result error:nil];
    }];
```
which will be super easy to test.

TL;DR I started to write tests for these results, but it would have been a lot of throwaway work for the next review.  Instead, I wrote all the tests that seemed like the best bang for my buck, and the next review tests diff will be easier to read, and not a wall o'green like this review.

File: packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m (hunk 7 above)
Comment (jmagman) [reply packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m:203]: Same comment as above, in GoogleSignIn 6.0 this became a callback `restorePreviousSignInWithCallback:` instead of a delegate.
https://github.com/flutter/plugins/blob/b548ae18646d1725a8e1f004c26bf42e4d362aa8/packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin.m#L103-L106
is handled in the delegate method:
https://github.com/flutter/plugins/blob/b548ae18646d1725a8e1f004c26bf42e4d362aa8/packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin.m#L208-L214
will become the easily testable:
```obj-c
    [GIDSignIn.sharedInstance restorePreviousSignInWithCallback:^(GIDGoogleUser *user, NSError *error) {
      [self didSignInForUser:user result:result withError:error];
    }];
```

File: packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m (hunk 8 above)
Comment (jmagman) [reply packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m:340]: The teardown calls `[self.mockSharedInstance stopMocking]`.
Only `OCMClassMock` objects with stubbed _class_ methods need `-stopMocking` called on them.  Stubbing on instance method calls `-stopMocking` when the instance is deallocated.  `mockSharedInstance` currently needs it because `[mockSharedInstance sharedInstance]` is mocked (though it won't be needed once I finish your injection suggestion).

**Comment by jmagman** [comment]
> The description says "Turn on code coverage", but I'm not seeing project changes other than deployment target.

Ah sorry I removed that in an intermediate commit since it's not being checked anywhere, so not worth the overhead in CI.  Fixed the description.

File: packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m (hunk 4 above)
Comment (jmagman) [reply packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m]: Done.

File: packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m (hunk 5 above)
Comment (jmagman) [reply packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m]: Done.

File: packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin.m (hunk 9 above)
Comment (jmagman) [reply packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin.m]: Done.

File: packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Private.h [hunk 10]
```diff
@@ -0,0 +1,13 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (jmagman) [inline packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Private.h:1]: Unfortunately I had to make this a public header to make it accessible to the RunnerTest.

Comment (jmagman) [reply packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Private.h:1]: Maybe I should rename to `FLTGoogleSignInPlugin_Test.h`

Comment (jmagman) [reply packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Private.h]: Just renamed it.

File: packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m (hunk 8 above)
Comment (stuartmorgan) [reply packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m]: Yep, I just can't read; I saw `OCMStub` and `sharedInstance` in the same line and thought this was stubbing another class method.

File: packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m (hunk 6 above)
Comment (stuartmorgan) [reply packages/google_sign_in/google_sign_in/example/ios/RunnerTests/GoogleSignInTests.m:71]: Makes sense, thanks!

File: packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Private.h (hunk 10 above)
Comment (stuartmorgan) [reply packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Private.h]: It would be kind of ugly, but couldn't the test do a path-based include of the private header instead of getting it via the module import?

Comment (stuartmorgan) [reply packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Private.h]: Alternately, you could do the thing that I almost always tell people *not* to do (in favor of making an internal header like you did) and re-declare the method you need at the top of the test file. I've never liked it, but it seems better than actually making the private header public in the module.

File: packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Test.h [hunk 11]
```diff
@@ -0,0 +1,13 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import <google_sign_in/FLTGoogleSignInPlugin.h>
+
+@class GIDSignIn;
+
+@interface FLTGoogleSignInPlugin ()
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Test.h:9]: I usually put a comment like "Methods exposed for testing" on these extensions, in case people who aren't familiar with that pattern are trying to understand what the header is for.

File: packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Test.h [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+- (instancetype)initWithSignIn:(GIDSignIn *)signIn NS_DESIGNATED_INITIALIZER;
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Test.h:11]: Maybe comment that this is explicitly for injecting the GIDSignIn for tests?

**Review by stuartmorgan — APPROVED** [review]
LGTM with the private header thing adrressed

File: packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Private.h (hunk 10 above)
Comment (jmagman) [reply packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Private.h]: How about a `Test` module so it needs to be explicitly imported?
```obj-c
@import google_sign_in;
@import google_sign_in.Test;
```

File: packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Test.h (hunk 11 above)
Comment (jmagman) [reply packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Test.h:12]: Done.

File: packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Test.h (hunk 12 above)
Comment (jmagman) [reply packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin_Test.h:15]: Done.

File: packages/google_sign_in/google_sign_in/pubspec.yaml [hunk 13]
```diff
@@ -3,7 +3,7 @@ description: Flutter plugin for Google Sign-In, a secure authentication system
   for signing in with a Google account on Android and iOS.
 repository: https://github.com/flutter/plugins/tree/master/packages/google_sign_in/google_sign_in
 issue_tracker: https://github.com/flutter/flutter/issues?q=is%3Aissue+is%3Aopen+label%3A%22p%3A+google_sign_in%22
-version: 5.0.4
+version: 5.0.5
```
Comment (jmagman) [inline packages/google_sign_in/google_sign_in/pubspec.yaml:6]: @stuartmorgan with the module change this seems different enough to warrant a patch bump.

File: packages/google_sign_in/google_sign_in/ios/google_sign_in.podspec [hunk 14]
```diff
@@ -12,8 +12,9 @@ Enables Google Sign-In in Flutter apps.
   s.license          = { :type => 'BSD', :file => '../LICENSE' }
   s.author           = { 'Flutter Team' => 'flutter-dev@googlegroups.com' }
   s.source           = { :http => 'https://github.com/flutter/plugins/tree/master/packages/google_sign_in' }
-  s.source_files = 'Classes/**/*'
+  s.source_files = 'Classes/**/*.{h,m}'
```
Comment (jmagman) [inline packages/google_sign_in/google_sign_in/ios/google_sign_in.podspec:15]: `modulemap` shouldn't be included in the source files.

File: packages/google_sign_in/google_sign_in/ios/Classes/google_sign_in-umbrella.h [hunk 15]
```diff
@@ -0,0 +1,10 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (jmagman) [inline packages/google_sign_in/google_sign_in/ios/Classes/google_sign_in-umbrella.h:1]: The file was previously generated by CocoaPods with the same name.  Now we need to manage it manually and check it into the repo because we're defining a module map, which takes us out of the magical auto module world.

File: packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin.modulemap [hunk 16]
```diff
@@ -0,0 +1,10 @@
+framework module google_sign_in {
+  umbrella header "google_sign_in-umbrella.h"
+
+  export *
+  module * { export * }
+
+  explicit module Test {
```
Comment (jmagman) [inline packages/google_sign_in/google_sign_in/ios/Classes/FLTGoogleSignInPlugin.modulemap:7]: This isn't a private module since there are some complications for Swift files not being able to import those.

Instead, this explicit module requires that `google_sign_in.Test` be imported to access FLTGoogleSignInPlugin_Test.  So `initWithSignIn:` won't be available in the normal `import google_sign_in` case, won't auto-complete, etc.

**Review by stuartmorgan — APPROVED** [review]
LGTM
